Dynavity/Dynavity/persistence/storage/CloudStorageManager.swift

```diff
@@ -0,0 +1,154 @@
+                    // if something is wrong, 'success' with empty value
+                    return callback(.success([]))
+                }
+                let futures = loaded.map { ref in
+                    Future<OnlineCanvasDTO, Never> { callback in
+                        let db = database.reference(withPath: "\(ref.userId)/self/\(ref.canvasId)")
+                        db.getData { _, snapshot in
+                            guard let value = snapshot.value,
+                                  let loaded = try? decoder.decode(OnlineCanvasDTO.self, from: value) else {
+                                return
+                            }
+                            callback(.success(loaded))
+                        }
+                    }
+                }
+                let canvases = MultiFutureSynchronizer(publishers: futures).blockForValue()
+                callback(.success(canvases))
+            }
+        }
+        let collabCanvases = FutureSynchronizer(publisher: collabFuture).blockForValue()
+        return (ownCanvases ?? []) + (collabCanvases ?? [])
+    }
+
+    func save(model canvas: OnlineCanvasDTO) throws {
+        guard let data = try? encoder.encode(canvas) else {
+            return
+        }
+        if userId == canvas.ownerId {
+            // save own canvas
+            let db = database.reference(withPath: "\(userId)/self/\(canvas.canvas.id)")
+            db.setValue(data)
+        } else {
+            // save other canvas
+            let ref = ExternalCanvasReference(userId: canvas.ownerId, canvasId: canvas.canvas.id.uuidString)
+            let db = database.reference(withPath: "\(ref.userId)/self/\(ref.canvasId)")
+            db.setValue(data)
+        }
+    }
+
+    func delete(model canvas: OnlineCanvasDTO) throws {
+        if userId == canvas.ownerId {
+            // delete own canvas
+            let db = database.reference(withPath: "\(userId)/self/\(canvas.canvas.id)")
+            db.removeValue()
+        } else {
+            // delete link to other canvas
+            let ref = ExternalCanvasReference(userId: canvas.ownerId, canvasId: canvas.canvas.id.uuidString)
+            let db = database.reference(withPath: "\(userId)/collab")
+            // get the whole array, filter, write back
+            db.getData { _, snapshot in
+                guard let value = snapshot.value,
+                      let loaded = try? decoder.decode([ExternalCanvasReference].self, from: value),
+                      let data = try? encoder.encode(loaded.filter { $0 != ref })else {
+                    return
+                }
+                db.setValue(data)
+            }
+        }
+    }
+}
+
+// MARK: Struct to help with decoding
+struct ExternalCanvasReference: Codable, Equatable {
+    let userId: String
+    let canvasId: String
+}
+
+// MARK: Helper to synchronize asynchronous code
```

Comment:
This converts all the asynchronous stuff from Firebase into synchronous code by blocking until a value is received.

It can move to its own package once the implementation is more stable.